### PR TITLE
Extend an expired key on re-purchase

### PR DESCRIPTION
**Pull request by hardlydifficult**
# Description

Fixes a bug where purchasing an expired key can result in still having an expired key.  Now on your second+ purchase the key will have an expiration date that is as good as or better than a brand new key.

I confirmed that if you undo the contract change the test I added fails.

# Issues

Fixes #
Refs #

# Checklist:

- [x] 1 PR, 1 purpose: my Pull Request applies to a single purpose
  - [ ] This PR only contains configuration changes (package.json, etc.)
  - [x] This PR only contains code changes (if configuration changes are required, do a separate PR first, then re-base)
- [ ] My code follows the style guidelines of this project, including naming conventions
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] If my code adds or changes components, I have written corresponding stories with Storybook
- [x] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: smart-contracts/contracts/mixins/MixinPurchase.sol [hunk 1]
```diff
@@ -84,21 +84,20 @@ contract MixinPurchase is
...
-    // We explicitly allow for greater amounts of ETH to allow 'donations'
-    _chargeAtLeast(netPrice);
-
     // Assign the key
     Key storage toKey = _getKeyFor(_recipient);
     uint previousExpiration = toKey.expirationTimestamp;
-    if (toKey.tokenId == 0) {
-      _assignNewTokenId(toKey);
-      _recordOwner(_recipient, toKey.tokenId);
+    if(toKey.expirationTimestamp >= block.timestamp) {
```
Comment (cellog) [inline smart-contracts/contracts/mixins/MixinPurchase.sol]:
```suggestion
    if (toKey.expirationTimestamp >= block.timestamp) {
```

Comment (hardlydifficult) [reply smart-contracts/contracts/mixins/MixinPurchase.sol]: thanks

File: smart-contracts/contracts/mixins/MixinPurchase.sol [hunk 2]
```diff
@@ -84,21 +84,20 @@ contract MixinPurchase is
...
-    // We explicitly allow for greater amounts of ETH to allow 'donations'
-    _chargeAtLeast(netPrice);
-
     // Assign the key
     Key storage toKey = _getKeyFor(_recipient);
     uint previousExpiration = toKey.expirationTimestamp;
-    if (toKey.tokenId == 0) {
-      _assignNewTokenId(toKey);
-      _recordOwner(_recipient, toKey.tokenId);
+    if (toKey.expirationTimestamp >= block.timestamp) {
```
Comment (nfurfaro) [inline smart-contracts/contracts/mixins/MixinPurchase.sol]: What if `toKey.expirationTimestamp <= block.timestamp` && `toKey.tokenId != 0` ?

Comment (nfurfaro) [inline smart-contracts/contracts/mixins/MixinPurchase.sol]: What if `toKey.expirationTimestamp <= block.timestamp` && `toKey.tokenId != 0` ?

File: smart-contracts/contracts/mixins/MixinPurchase.sol [hunk 3]
```diff
@@ -84,21 +84,20 @@ contract MixinPurchase is
...
+      // This is an existing owner trying to extend their key
+      toKey.expirationTimestamp = previousExpiration.add(expirationDuration);
+    } else {
+      if (toKey.tokenId == 0) {
+        _assignNewTokenId(toKey);
+        _recordOwner(_recipient, toKey.tokenId);
+      }
       // SafeAdd is not required here since expirationDuration is capped to a tiny value
       // (relative to the size of a uint)
       toKey.expirationTimestamp = block.timestamp + expirationDuration;
```
Comment (nfurfaro) [inline smart-contracts/contracts/mixins/MixinPurchase.sol]: won't this line be executed even if line #92 is also executed?

Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinPurchase.sol]: it seems like we want either line 92 or line 100 to be executed...not both.

Comment (hardlydifficult) [reply smart-contracts/contracts/mixins/MixinPurchase.sol]: Na, we are nested inside an `else` block here.

**Comment by hardlydifficult** [comment]
Not sure why this caused some transfer tests to fail... investigating..

**Comment by hardlydifficult** [comment]
oh, this change conflicts with MixinTransfer.sol line 91.  I'll try a different approach.

File: smart-contracts/contracts/mixins/MixinPurchase.sol (hunk 3 above)
Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinPurchase.sol]: yup.

**Comment by nfurfaro** [comment]
might need something like `if toKey.expirationTimestamp < block.timestamp && toKey.tokenId != 0`

**Comment by hardlydifficult** [comment]
k should be good to go - could you take another look?

**Pull request merged**